src/Badges.sol

```diff
@@ -143,6 +143,28 @@ contract Badges is
     return owner_;
   }
 
+  function deactivateBadge(
+    uint256 _raftTokenId,
+    uint256 _badgeId,
+    string memory _reason
+  ) external {
+    require(specDataHolder.getRaftOwner(_raftTokenId) == msg.sender, "createSpec: unauthorized");
+    // check the deactivation permissions
+    badgeIdToDeactivationReason[_badgeId] = _reason;
+  }
+
+  function reactivateBadge(uint256 _raftTokenId, uint256 _badgeId) external {
+    require(specDataHolder.getRaftOwner(_raftTokenId) == msg.sender, "createSpec: unauthorized");
+    delete badgeIdToDeactivationReason[_badgeId];
+  }
+
+  function isBadgeValid(uint256 _badgeId, uint256 timestamp) external view returns (bool) {
+    // if solidity were friendlier we could do `badgeIdToDeactivationReason[badgeId] == ""`
+    bool isntDeactivated = keccak256(abi.encode(badgeIdToDeactivationReason[_badgeId])) == keccak256("");
+    bool isntExpired = timestamp < block.timestamp;
```

Comment:
1 - block.timestamp captures the timestamp of when the block was created, right? what's the rationale for using this?
2 - we need the badge expiration already captured on-chain to compute this, no?